[Fix] Convert nodes to RST nodes before comparing

eddyerburgh opened the pull request:
- Convert nodes in `contains` functions to RST before comparing 

- Fixes #1422

ljharb commented on packages/enzyme/src/ReactWrapper.js:
```diff
@@ -446,7 +454,10 @@ class ReactWrapper {
    */
   containsAnyMatchingElements(nodes) {
-    return Array.isArray(nodes) && nodes.some(node => this.containsMatchingElement(node));
+    const rstNodes = nodes.map(getAdapter().elementToNode);
```
this change makes `.map` be called before `Array.isArray` is called to ensure it's an array.

Let's change this to `if (!Array.isArray(nodes)) { return false; }` at the top of the function, and then we can do the entire check in a single `some`

ljharb commented on packages/enzyme/src/ReactWrapper.js:
```diff
@@ -374,9 +375,13 @@ class ReactWrapper {
+    const rstNodeOrNodes = isArray
+      ? nodeOrNodes.map(getAdapter().elementToNode)
+      : getAdapter().elementToNode(nodeOrNodes);
     const predicate = Array.isArray(nodeOrNodes)
```
this line can use `isArray` now, I think?

ljharb commented on packages/enzyme/src/ReactWrapper.js:
```diff
@@ -374,9 +375,13 @@ class ReactWrapper {
     const predicate = Array.isArray(nodeOrNodes)
-      ? other => containsChildrenSubArray(nodeEqual, other, nodeOrNodes)
-      : other => nodeEqual(nodeOrNodes, other);
+      ? other => containsChildrenSubArray(nodeEqual, other, rstNodeOrNodes)
```
probably at this point, rather than two paired multiline ternaries, this would be better as an `if`/`else` branch.

ljharb commented on packages/enzyme/src/ReactWrapper.js:
```diff
@@ -358,7 +358,8 @@ class ReactWrapper {
   matchesElement(node) {
-    return this.single('matchesElement', () => nodeMatches(node, this.getNodeInternal(), (a, b) => a <= b));
+    const rstNode = getAdapter().elementToNode(node);
+    return this.single('matchesElement', () => nodeMatches(rstNode, this.getNodeInternal(), (a, b) => a <= b));
```
let's do the elementToNode conversion here inside the arrow function, so that it happens after `this.single` succeeds.

ljharb commented:
@eddyerburgh is this change not needed in `shallow` as well?

eddyerburgh commented:
@ljharb Yes, you're right. I've only been implementing mount so haven't looked into shallow. I'll add the changes now

ljharb commented on packages/enzyme/src/ReactWrapper.js:
```diff
@@ -374,9 +377,16 @@ class ReactWrapper {
+      ? other => containsChildrenSubArray(
+        nodeEqual,
+        other,
+        nodeOrNodes.map(adapter.elementToNode),
```
Let’s make this use an arrow function, so adapters only get the single argument

ljharb replied:
I also see this is how it already is in https://github.com/airbnb/enzyme/pull/1423/files#diff-fa27c82ee4fde075bae6173848e17c82R448; let’s change both

eddyerburgh replied:
> Let’s make this use an arrow function, so adapters only get the single argument
 
Do you mean the `nodeOrNodes` `map` callback should be an arrow function?

ljharb replied:
yep! like `.map(x => adapter.elementToNode(x))`

eddyerburgh replied:
Done 🙂

eddyerburgh commented:
It was already implemented for contains in ShallowWrapper, so I refactored `contains` in ReactWrapper. 

I also realized `containsAllMatchingElements` and `containsAnyMatchingElements` call `containsMatchingElement`, so I removed the unncessary calls to `containsMatchingElement`.

Finally, while I'm here, in `ReactWrapper.contains` there's no check that the node/nodes is a React element. Should I add a test and write a check, like there is [in `ShallowWrapper.contains`](https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ShallowWrapper.js#L441)?

ljharb commented:
> I also realized containsAllMatchingElements and containsAnyMatchingElements call containsMatchingElement, so I removed the unncessary calls to containsMatchingElement.

This was done intentionally, so there couldn’t be any drift between how these three methods worked - please keep them linked.

ljharb commented:
In `mount`, someone might pass a DOM node; I’m not sure what the best solution is there.

eddyerburgh commented:
>  so I removed the unncessary calls to `containsMatchingElement`

Sorry, I meant that I removed the unnecessary calls to `elementToNode`. They are all still linked.

ljharb commented:
gotcha, sounds good then :-)

eddyerburgh commented:
Thanks @ljharb . Is there anything else you need me to do to get this merged?

Also, do you know when this will be released?

ljharb commented:
@eddyerburgh at the moment, you're good; i'm waiting for other collaborators to review it first.

After it's merged, it will be released whenever we're ready to publish the next one :-) hopefully soon.

eddyerburgh commented:
@collaborators, could we get this reviewed and merged?

ljharb commented:
The following packages are now released:
 - [enzyme v3.4.0](https://www.npmjs.com/package/enzyme)
 - [enzyme-adapter-react-16 v1.2.0](https://www.npmjs.com/package/enzyme-adapter-react-16)
 - [enzyme-adapter-react-16.3 v1.0.0](https://www.npmjs.com/package/enzyme-adapter-react-16.3)
 - [enzyme-adapter-react-16.2 v1.0.0](https://www.npmjs.com/package/enzyme-adapter-react-16.2)
 - [enzyme-adapter-react-16.1 v1.0.0](https://www.npmjs.com/package/enzyme-adapter-react-16.1)
 - [enzyme-adapter-react-15 v1.0.6](https://www.npmjs.com/package/enzyme-adapter-react-15)
 - [enzyme-adapter-react-15.4 v1.0.6](https://www.npmjs.com/package/enzyme-adapter-react-15.4)
 - [enzyme-adapter-react-14 v1.0.6](https://www.npmjs.com/package/enzyme-adapter-react-14)
 - [enzyme-adapter-react-13 v1.0.4](https://www.npmjs.com/package/enzyme-adapter-react-13)
 - [enzyme-adapter-utils v1.5.0](https://www.npmjs.com/package/enzyme-adapter-utils)
